File: 16-detect-cycle/solution.go
// Detect Cycle in Linked List (Floyd's Cycle Detection)
// Time: O(n), Space: O(1)

package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

type Solution struct{}

func (s *Solution) HasCycle(head *ListNode) bool {
	if head == nil || head.Next == nil {
		return false
	}
	
	slow := head
	fast := head
	
	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
		if slow == fast {
			return true
		}
	}
	
	return false
}

func (s *Solution) DetectCycle(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return nil
	}
	
	slow := head
	fast := head
	
	// Find meeting point
	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
		if slow == fast {
			break
		}
	}
	
	if fast == nil || fast.Next == nil {
		return nil
	}
	
	// Find cycle start
	slow = head
	for slow != fast {
		slow = slow.Next
		fast = fast.Next
	}
	
	return slow
}

func main() {
	head := &ListNode{Val: 1}
	node2 := &ListNode{Val: 2}
	node3 := &ListNode{Val: 3}
	node4 := &ListNode{Val: 4}
	head.Next = node2
	node2.Next = node3
	node3.Next = node4
	node4.Next = node2
	
	sol := &Solution{}
	fmt.Println(sol.HasCycle(head))  // true
	cycleNode := sol.DetectCycle(head)
	if cycleNode != nil {
		fmt.Println(cycleNode.Val)  // 2
	} else {
		fmt.Println("nil")
	}
}

File: 16-detect-cycle/solution.java
// Detect Cycle in Linked List (Floyd's Cycle Detection)
// Time: O(n), Space: O(1)

class ListNode {
    int val;
    ListNode next;
    
    ListNode(int val) { this.val = val; }
}

class Solution {
    public boolean hasCycle(ListNode head) {
        if (head == null || head.next == null) {
            return false;
        }
        
        ListNode slow = head;
        ListNode fast = head;
        
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
            if (slow == fast) {
                return true;
            }
        }
        
        return false;
    }
    
    public ListNode detectCycle(ListNode head) {
        if (head == null || head.next == null) {
            return null;
        }
        
        ListNode slow = head;
        ListNode fast = head;
        
        // Find meeting point
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
            if (slow == fast) {
                break;
            }
        }
        
        if (fast == null || fast.next == null) {
            return null;
        }
        
        // Find cycle start
        slow = head;
        while (slow != fast) {
            slow = slow.next;
            fast = fast.next;
        }
        
        return slow;
    }
    
    public static void main(String[] args) {
        ListNode head = new ListNode(1);
        ListNode node2 = new ListNode(2);
        ListNode node3 = new ListNode(3);
        ListNode node4 = new ListNode(4);
        head.next = node2;
        node2.next = node3;
        node3.next = node4;
        node4.next = node2;
        
        Solution sol = new Solution();
        System.out.println(sol.hasCycle(head));  // true
        ListNode cycleNode = sol.detectCycle(head);
        System.out.println(cycleNode != null ? cycleNode.val : "null");  // 2
    }
}
